io-engine/src/grpc/v1/stats.rs

```diff
@@ -0,0 +1,200 @@
+use crate::{
+    core::lock::ResourceLockManager,
+    grpc::{
+        rpc_submit,
+        v1::{pool::PoolService, replica::ReplicaService},
+        GrpcClientContext,
+        GrpcResult,
+        ReadLock,
+        Serializer,
+    },
+    lvs::Lvs,
+};
+use futures::FutureExt;
+use io_engine_api::v1::stats::*;
+use std::fmt::Debug;
+use tonic::{Request, Response, Status};
+
+use crate::core::{BlockDeviceIoStats, CoreError, UntypedBdev};
+use ::function_name::named;
+use std::panic::AssertUnwindSafe;
+
+/// RPC service for Resource IoStats.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct StatsService {
+    name: String,
+    client_context:
+        std::sync::Arc<tokio::sync::RwLock<Option<GrpcClientContext>>>,
+    pool_svc: PoolService,
+    replica_svc: ReplicaService,
+}
+
+#[async_trait::async_trait]
+impl<F, T> Serializer<F, T> for StatsService
+where
+    T: Send + 'static,
+    F: core::future::Future<Output = Result<T, Status>> + Send + 'static,
+{
+    async fn locked(&self, ctx: GrpcClientContext, f: F) -> Result<T, Status> {
+        // Takes read lock of Pool and Replica service.
+        let pool_context = self.pool_svc.client_context.read().await;
+        let replica_context = self.replica_svc.client_context.read().await;
+
+        if let Some(c) = pool_context.as_ref() {
+            warn!("{}: gRPC method timed out, args: {}", c.id, c.args);
+        }
+        if let Some(c) = replica_context.as_ref() {
+            warn!("{}: gRPC method timed out, args: {}", c.id, c.args);
+        }
```

Comment:
If you really need the context, you could modify the read to return (Guard,Option<Context>)

```suggestion
        let _pool = self.pool_svc.read().await;
        let _replica = self.replica_svc.read().await;
```

Reply:
dont need client context.